cmd/repo-query/repo-query.go

```diff
@@ -0,0 +1,353 @@
+	// Remove the "https://github.com/" prefix from the URL
+	url = strings.TrimPrefix(url, "https://github.com/")
+
+	// Split the remaining URL path into segments
+	segments := strings.Split(url, "/")
+
+	// The first segment is the owner, and the second segment is the repository name
+	if len(segments) >= 2 {
+		owner = segments[0]
+		repo = segments[1]
+	} else {
+		// URL is not in the expected format
+		owner = ""
+		repo = ""
+	}
+
+	return owner, repo
+}
+
+func run(opts *Options) error {
+	// get repo name and owner name from URL
+	owner, repo := getOwnerAndRepo(opts.URL)
+
+	fmt.Printf("Checking if %s/%s is indexed by us...⏳\n", owner, repo)
+	resp, err := http.Get(fmt.Sprintf("%s/collection?owner=%s&name=%s&branch=%s", repoQueryURL, owner, repo, opts.branch))
+	if err != nil {
+		return err
+	}
+
+	// not found or ok or error
+	if resp.StatusCode == http.StatusNotFound {
+		fmt.Println("Repo not found❗")
+		fmt.Println("Indexing repo...⏳")
+		err := indexRepo(owner, repo, opts.branch)
+		if err != nil {
+			return err
+		}
+
+		for {
+			fmt.Printf("\nWant to ask a question about %s/%s?\n", owner, repo)
+			fmt.Printf("> ")
+			// read input
+			scanner := bufio.NewScanner(os.Stdin)
+			if scanner.Scan() {
+				input := scanner.Text()
+				err := askQuestion(input, owner, repo, opts.branch)
+				if err != nil {
+					return err
+				}
+			}
+		}
+	} else if resp.StatusCode == http.StatusOK {
+		fmt.Println("Repo found ✅")
+
+		// this should be an infinite loop
+		for {
+			fmt.Printf("\nWant to ask a question about %s/%s?\n", owner, repo)
+			fmt.Printf("> ")
+			// read input
+			scanner := bufio.NewScanner(os.Stdin)
+			if scanner.Scan() {
+				input := scanner.Text()
+				err := askQuestion(input, owner, repo, opts.branch)
+				if err != nil {
+					return err
+				}
+			}
+		}
+	} else {
+		body, err := io.ReadAll(resp.Body)
+		if err != nil {
+			return err
+		}
+
+		fmt.Printf("An error occurred: %v\n", string(body))
+		return errors.New("error while checking if repo is indexed")
+	}
+}
+
+type indexPostRequest struct {
+	Owner  string `json:"owner"`
+	Name   string `json:"name"`
+	Branch string `json:"branch"`
+}
+
+func indexRepo(owner string, repo string, branch string) error {
+	indexPostReq := &indexPostRequest{
+		Owner:  owner,
+		Name:   repo,
+		Branch: branch,
+	}
+
+	indexPostJSON, err := json.Marshal(indexPostReq)
+	if err != nil {
+		return err
+	}
+
+	responseBody := bytes.NewBuffer(indexPostJSON)
+	resp, err := http.Post(fmt.Sprintf("%s/embed", repoQueryURL), "application/json", responseBody)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		body, err := io.ReadAll(resp.Body)
+		if err != nil {
+			return err
+		}
+
+		fmt.Printf("An error occurred: %v\n", string(body))
+		return errors.New("error while indexing repository")
+	}
+
+	// listen for SSEs and send data,event pairs to processIndexChunk
+	reader := bufio.NewReader(resp.Body)
+	for {
```

Comment:
This whole chunk from 175 through 209 seems really brittle.

But maybe I'm not understanding it fully, let me see if I have it right: essentially, we're anticipating getting server side events where:

- EOF from the response body indicates the server events are done
- new lines can be thrown away and are essentially empty events from the server.
- Anything that has `event: ` indicates a new event from the server and we should process the string chunks up to (and including the newline).
  - A empty newline string in this chunk indicates that event is done (?)

Maybe this all just needs some more inline documentation since it's not immediately clear _how_ this is supposed to work or how it's supposed to interact with the server. Even better, we could write a clear "handler" that has individual methods for each scenario. But that may be overkill: regardless, we should harden this since I anticipate this going stale really quickly.

Reply:
Yup, I had to handle some of these cases (empty events, inconsistent responses etc.)   here too - https://github.com/open-sauced/ai/pull/226/files#diff-34c2ff6688d75e301c74751cff9760478c59b33e97c949f5b05dfe1f273517d2R94. I just had the context already so I pasted things in here. Will document this.